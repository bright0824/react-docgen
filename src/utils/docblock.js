/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

/**
 * Helper functions to work with docblock comments.
 */

var DOCLET_PATTERN = /^@(\w+)(?:$|\s((?:[^](?!^@\w))*))/gmi;

function parseDocblock(str) {
  var lines = str.split('\n');
  for (var i = 0, l = lines.length; i < l; i++) {
    lines[i] = lines[i].replace(/^\s*\*\s?/, '');
  }
  return lines.join('\n').trim();
}

/**
 * Given a path, this function returns the closest preceding docblock if it
 * exists.
 */
export function getDocblock(path: NodePath): ?string {
  if (path.node.comments) {
    var comments = path.node.comments.filter(function(comment) {
      return comment.leading &&
        comment.type === 'CommentBlock' &&
        comment.value.indexOf('*\n') === 0;
    });
    if (comments.length > 0) {
      return parseDocblock(comments[comments.length - 1].value);
    }
  }
  return null;
}

/**
 * Given a string, this functions returns an object with doclet names as keys
 * and their "content" as values.
 */
export function getDoclets(str: string): Object {
  var doclets = Object.create(null);
  var match = DOCLET_PATTERN.exec(str);

  for (; match; match = DOCLET_PATTERN.exec(str)) {
    doclets[match[1]] = match[2] || true;
  }

  return doclets;
}
