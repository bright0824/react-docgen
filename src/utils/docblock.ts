/**
 * Helper functions to work with docblock comments.
 */

import { CommentKind } from 'ast-types/gen/kinds';
import type { NodePath } from 'ast-types/lib/node-path';

const DOCLET_PATTERN = /^@(\w+)(?:$|\s((?:[^](?!^@\w))*))/gim;

function parseDocblock(str: string): string {
  // Does not use \s in the regex as this would match also \n and conflicts
  // with windows line endings.
  return str.replace(/^[ \t]*\*[ \t]?/gm, '').trim();
}

const DOCBLOCK_HEADER = /^\*\s/;

/**
 * Given a path, this function returns the closest preceding docblock if it
 * exists.
 */
export function getDocblock(path: NodePath, trailing = false): string | null {
  let comments: CommentKind[] = [];
  if (trailing && path.node.trailingComments) {
    comments = path.node.trailingComments.filter(
      comment =>
        comment.type === 'CommentBlock' && DOCBLOCK_HEADER.test(comment.value),
    );
  } else if (path.node.leadingComments) {
    comments = path.node.leadingComments.filter(
      comment =>
        comment.type === 'CommentBlock' && DOCBLOCK_HEADER.test(comment.value),
    );
  } else if (path.node.comments) {
    comments = path.node.comments.filter(
      comment =>
        comment.leading &&
        comment.type === 'CommentBlock' &&
        DOCBLOCK_HEADER.test(comment.value),
    );
  }

  if (comments.length > 0) {
    return parseDocblock(comments[comments.length - 1].value);
  }
  return null;
}

/**
 * Given a string, this functions returns an object with doclet names as keys
 * and their "content" as values.
 */
export function getDoclets(str: string): Record<string, unknown> {
  const doclets = Object.create(null);
  let match = DOCLET_PATTERN.exec(str);

  for (; match; match = DOCLET_PATTERN.exec(str)) {
    doclets[match[1]] = match[2] || true;
  }

  return doclets;
}
