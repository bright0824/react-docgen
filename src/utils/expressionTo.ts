/*eslint no-loop-func: 0, no-use-before-define: 0*/

import { namedTypes as t } from 'ast-types';
import resolveToValue from './resolveToValue';
import type { Importer } from '../parse';
import type { NodePath } from 'ast-types/lib/node-path';

/**
 * Splits a MemberExpression or CallExpression into parts.
 * E.g. foo.bar.baz becomes ['foo', 'bar', 'baz']
 */
function toArray(path: NodePath, importer: Importer): string[] {
  const parts = [path];
  let result: string[] = [];

  while (parts.length > 0) {
    path = parts.shift() as NodePath;
    const node = path.node;
    if (t.CallExpression.check(node)) {
      parts.push(path.get('callee'));
      continue;
    } else if (t.MemberExpression.check(node)) {
      parts.push(path.get('object'));
      if (node.computed) {
        const resolvedPath = resolveToValue(path.get('property'), importer);
        if (resolvedPath !== undefined) {
          result = result.concat(toArray(resolvedPath, importer));
        } else {
          result.push('<computed>');
        }
      } else {
        // @ts-ignore
        result.push(node.property.name);
      }
      continue;
    } else if (t.Identifier.check(node)) {
      result.push(node.name);
      continue;
    } else if (t.Literal.check(node)) {
      // @ts-ignore
      result.push(node.raw);
      continue;
    } else if (t.FunctionExpression.check(node)) {
      result.push('<function>');
      continue;
    } else if (t.ThisExpression.check(node)) {
      result.push('this');
      continue;
    } else if (t.ObjectExpression.check(node)) {
      const properties = path.get('properties').map(function (property) {
        if (t.SpreadElement.check(property.node)) {
          return `...${toString(property.get('argument'), importer)}`;
        } else {
          return (
            toString(property.get('key'), importer) +
            ': ' +
            toString(property.get('value'), importer)
          );
        }
      });
      result.push('{' + properties.join(', ') + '}');
      continue;
    } else if (t.ArrayExpression.check(node)) {
      result.push(
        '[' +
          path
            .get('elements')
            .map(function (el) {
              return toString(el, importer);
            })
            .join(', ') +
          ']',
      );
      continue;
    }
  }

  return result.reverse();
}

/**
 * Creates a string representation of a member expression.
 */
function toString(path: NodePath, importer: Importer): string {
  return toArray(path, importer).join('.');
}

export { toString as String, toArray as Array };
