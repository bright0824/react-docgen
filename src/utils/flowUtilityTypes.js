/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */
import recast from 'recast';

const {
  types: { namedTypes: types },
} = recast;

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(path: NodePath): boolean {
  if (types.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return Boolean(idPath) && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  return path.get('typeParameters', 'params', 0);
}
