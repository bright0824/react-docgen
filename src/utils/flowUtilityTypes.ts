import { namedTypes as t } from 'ast-types';
import type { NodePath } from 'ast-types/lib/node-path';

const supportedUtilityTypes = new Set(['$Exact', '$ReadOnly']);

/**
 * See `supportedUtilityTypes` for which types are supported and
 * https://flow.org/en/docs/types/utilities/ for which types are available.
 */
export function isSupportedUtilityType(path: NodePath): boolean {
  if (t.GenericTypeAnnotation.check(path.node)) {
    const idPath = path.get('id');
    return !!idPath && supportedUtilityTypes.has(idPath.node.name);
  }
  return false;
}

/**
 * Unwraps well known utility types. For example:
 *
 *   $ReadOnly<T> => T
 */
export function unwrapUtilityType(path: NodePath): NodePath {
  while (isSupportedUtilityType(path)) {
    path = path.get('typeParameters', 'params', 0);
  }

  return path;
}
