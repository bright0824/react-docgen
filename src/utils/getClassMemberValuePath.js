/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import getNameOrValue from './getNameOrValue';
import recast from 'recast';

const {
  types: { namedTypes: types },
} = recast;

export default function getClassMemberValuePath(
  classDefinition: NodePath,
  memberName: string,
): ?NodePath {
  // Fortunately it seems like that all members of a class body, be it
  // ClassProperty or MethodDefinition, have the same structure: They have a
  // "key" and a "value"
  return classDefinition
    .get('body', 'body')
    .filter(
      memberPath =>
        (!memberPath.node.computed ||
          types.Literal.check(memberPath.node.key)) &&
        !types.PrivateName.check(memberPath.node.key) &&
        getNameOrValue(memberPath.get('key')) === memberName &&
        memberPath.node.kind !== 'set',
    )
    .map(memberPath => memberPath.get('value'))[0];
}
