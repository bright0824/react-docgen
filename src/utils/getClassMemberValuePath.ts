import type { NodePath } from '@babel/traverse';
import type {
  ClassDeclaration,
  ClassExpression,
  ClassMethod,
  ClassProperty,
  Expression,
} from '@babel/types';
import getNameOrValue from './getNameOrValue';

export default function getClassMemberValuePath(
  classDefinition: NodePath<ClassDeclaration | ClassExpression>,
  memberName: string,
): NodePath<ClassMethod | Expression> | null {
  // Fortunately it seems like that all members of a class body, be it
  // ClassProperty or ClassMethod, have the same structure: They have a
  // "key" and a "value"
  return (
    classDefinition
      .get('body')
      .get('body')
      .filter(memberPath => {
        if (
          (memberPath.isClassMethod() && memberPath.node.kind !== 'set') ||
          memberPath.isClassProperty()
        ) {
          const key = (memberPath as NodePath<ClassMethod | ClassProperty>).get(
            'key',
          );
          return (
            (!memberPath.node.computed || key.isLiteral()) &&
            getNameOrValue(key) === memberName
          );
        }

        return false;
      }) //TODO ClassMethod does not have value
      .map(memberPath =>
        memberPath.isClassMethod()
          ? memberPath
          : (memberPath.get('value') as NodePath<Expression>),
      )[0] || null
  );
}
