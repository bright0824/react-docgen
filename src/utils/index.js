/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 *
 */

import * as docblock from './docblock';

export {
  docblock,
};

export {default as getClassMemberValuePath} from './getClassMemberValuePath';
export {default as getMemberExpressionRoot} from './getMemberExpressionRoot';
export {default as getMembers} from './getMembers';
export {default as getMemberValuePath} from './getMemberValuePath';
export {default as getNameOrValue} from './getNameOrValue';
export {default as getPropertName} from './getPropertyName';
export {default as getPropertyValuePath} from './getPropertyValuePath';
export {default as getPropType} from './getPropType';
export {default as isExportsOrModuleAssignment} from './isExportsOrModuleAssignment';
export {default as isReactComponentClass} from './isReactComponentClass.js';
export {default as isReactCreateClassCall} from './isReactCreateClassCall';
export {default as isReactModuleName} from './isReactModuleName';
export {default as match} from './match';
export {default as normalizeClassDefiniton} from './normalizeClassDefinition';
export {default as printValue} from './printValue';
export {default as resolveExportDeclaration} from './resolveExportDeclaration';
export {default as resolveToModule} from './resolveToModule';
export {default as resolveToValue} from './resolveToValue';
