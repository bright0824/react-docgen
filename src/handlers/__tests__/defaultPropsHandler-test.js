/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

jest.mock('../../Documentation');

import { parse } from '../../../tests/utils';

describe('defaultPropsHandler', () => {
  let documentation;
  let defaultPropsHandler;

  beforeEach(() => {
    documentation = new (require('../../Documentation'))();
    defaultPropsHandler = require('../defaultPropsHandler').default;
  });

  describe('ObjectExpression', () => {
    it('should find prop default values that are literals', () => {
      const src = `
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              bar: 42,
              baz: ["foo", "bar"],
              abc: {xyz: abc.def, 123: 42}
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('handles computed properties', () => {
      const src = `
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [bar]: 42,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('ignores complex computed properties', () => {
      const src = `
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [() => {}]: 42,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves local spreads', () => {
      const src = `
        const other = { bar: "foo" };

        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              ...other,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('ClassDeclaration with static defaultProps', () => {
    it('should find prop default values that are literals', () => {
      const src = `
        class Foo {
          static defaultProps = {
            foo: "bar",
            bar: 42,
            baz: ["foo", "bar"],
            abc: {xyz: abc.def, 123: 42}
          };
        }
      `;
      defaultPropsHandler(documentation, parse(src).get('body', 0));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should resolve local spreads', () => {
      const src = `
        const other = { bar: "foo" };

        class Foo {
          static defaultProps = {
            foo: "bar",
            ...other
          };
        }
      `;
      defaultPropsHandler(documentation, parse(src).get('body', 1));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should find prop default values that are imported variables', () => {
      const src = `
        import ImportedComponent from './ImportedComponent';

        class Foo {
          static defaultProps = {
            foo: ImportedComponent,
          };
        }
      `;
      defaultPropsHandler(documentation, parse(src).get('body', 1));
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('ClassExpression with static defaultProps', () => {
    it('should find prop default values that are literals', () => {
      const src = `
        var Bar = class {
          static defaultProps = {
            foo: "bar",
            bar: 42,
            baz: ["foo", "bar"],
            abc: {xyz: abc.def, 123: 42}
          };
      }`;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'declarations', 0, 'init'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  it('should only consider Property nodes, not e.g. spread properties', () => {
    const src = `
      ({
        getDefaultProps: function() {
          return {
            ...Foo.bar,
            bar: 42,
          };
        }
      })
    `;
    const definition = parse(src).get('body', 0, 'expression');
    expect(() => defaultPropsHandler(documentation, definition)).not.toThrow();
    expect(documentation.descriptors).toMatchSnapshot();
  });

  describe('Functional components with default params', () => {
    it('should find default props that are literals', () => {
      const src = `
        ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc = {xyz: abc.def, 123: 42}
        }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should override with defaultProps if available', () => {
      const src = `
        var Foo = ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc = 'test'
        }) => <div />
        Foo.defaultProps = { abc: {xyz: abc.def, 123: 42} };
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'declarations', 0, 'init'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves local spreads', () => {
      const src = `
        const other = { bar: "foo" };
        var Foo = (props) => <div />
        Foo.defaultProps = { foo: "bar", ...other };
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'declarations', 0, 'init'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should work with aliases', () => {
      const src = `
        ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc: defg = {xyz: abc.def, 123: 42}
        }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should find prop default values that are imported variables', () => {
      const src = `
        import ImportedComponent from './ImportedComponent';

        ({
          foo = ImportedComponent,
        }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should work with no defaults', () => {
      const src = `
        ({ foo }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('forwardRef', () => {
    it('resolves default props in the parameters', () => {
      const src = `
        import React from 'react';
        React.forwardRef(({ foo = 'bar' }, ref) => <div ref={ref}>{foo}</div>);
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves defaultProps', () => {
      const src = `
        import React from 'react';
        const Component = React.forwardRef(({ foo }, ref) => <div ref={ref}>{foo}</div>);
        Component.defaultProps = { foo: 'baz' };
      `;
      defaultPropsHandler(documentation, parse(src).get('body', 1));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves when the function is not inline', () => {
      const src = `
        import React from 'react';
        const ComponentImpl = ({ foo = 'bar' }, ref) => <div ref={ref}>{foo}</div>;
        React.forwardRef(ComponentImpl);
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 2, 'expression'),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });
});
