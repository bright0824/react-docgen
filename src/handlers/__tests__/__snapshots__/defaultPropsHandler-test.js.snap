// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should find prop default values that are imported variables 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": true,
      "value": "ImportedComponent",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should find prop default values that are literals 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should resolve local spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassExpression with static defaultProps should find prop default values that are literals 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params resolves local spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should find default props that are literals 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should find prop default values that are imported variables 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": true,
      "value": "ImportedComponent",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should override with defaultProps if available 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should work with aliases 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should work with no defaults 1`] = `Object {}`;

exports[`defaultPropsHandler ObjectExpression handles computed properties 1`] = `
Object {
  "@computed#bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression ignores complex computed properties 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression resolves local spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression should find prop default values that are literals 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves default props in the parameters 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "'bar'",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves defaultProps 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "'baz'",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves when the function is not inline 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "'bar'",
    },
  },
}
`;

exports[`defaultPropsHandler should only consider Property nodes, not e.g. spread properties 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
}
`;
